src/Open3D/GUI/SceneWidget.cpp

```diff
@@ -0,0 +1,895 @@
+    }
+
+    void SetDirectionalLight(
+            visualization::LightHandle dirLight,
+            std::function<void(const Eigen::Vector3f&)> onChanged) {
+        lightDir_->SetDirectionalLight(dirLight, onChanged);
+    }
+
+    void SetSkyboxHandle(visualization::SkyboxHandle skybox, bool isOn) {
+        ibl_->SetSkyboxHandle(skybox, isOn);
+    }
+
+    void SetModel(visualization::GeometryHandle axes,
+                  const std::vector<visualization::GeometryHandle>& objects) {
+        model_->SetModel(axes, objects);
+    }
+
+    SceneWidget::Controls GetControls() const {
+        if (current_ == fps_.get()) {
+            return SceneWidget::Controls::FPS;
+        } else if (current_ == lightDir_.get()) {
+            return SceneWidget::Controls::ROTATE_SUN;
+        } else if (current_ == ibl_.get()) {
+            return SceneWidget::Controls::ROTATE_IBL;
+        } else if (current_ == model_.get()) {
+            return SceneWidget::Controls::ROTATE_MODEL;
+        } else {
+            return SceneWidget::Controls::ROTATE_OBJ;
+        }
+    }
+
+    void SetControls(SceneWidget::Controls mode) {
+        switch (mode) {
+            case SceneWidget::Controls::ROTATE_OBJ:
+                current_ = rotate_.get();
+                break;
+            case SceneWidget::Controls::FPS:
+                current_ = fps_.get();
+                break;
+            case SceneWidget::Controls::ROTATE_SUN:
+                current_ = lightDir_.get();
+                break;
+            case SceneWidget::Controls::ROTATE_IBL:
+                current_ = ibl_.get();
+                break;
+            case SceneWidget::Controls::ROTATE_MODEL:
+                current_ = model_.get();
+                break;
+        }
+    }
+
+    void Mouse(const MouseEvent& e) {
+        if (current_ == rotate_.get()) {
+            if (e.type == MouseEvent::Type::BUTTON_DOWN &&
+                (e.button.button == MouseButton::MIDDLE ||
+                 e.modifiers == int(KeyModifier::ALT))) {
+                override_ = lightDir_.get();
+            }
+        }
+
+        if (override_) {
+            override_->Mouse(e);
+        } else if (current_) {
+            current_->Mouse(e);
+        }
+
+        if (override_ && e.type == MouseEvent::Type::BUTTON_UP) {
+            override_ = nullptr;
+        }
+    }
+
+    void Key(const KeyEvent& e) {
+        if (current_) {
+            current_->Key(e);
+        }
+    }
+
+    Widget::DrawResult Tick(const TickEvent& e) {
+        if (current_) {
+            if (current_->Tick(e)) {
+                return Widget::DrawResult::REDRAW;
+            }
+        }
+        return Widget::DrawResult::NONE;
+    }
+
+private:
+    std::unique_ptr<RotateCameraInteractor> rotate_;
+    std::unique_ptr<FPSInteractor> fps_;
+    std::unique_ptr<RotateSunInteractor> lightDir_;
+    std::unique_ptr<RotateIBLInteractor> ibl_;
+    std::unique_ptr<RotateModelInteractor> model_;
```

Comment:
FPS -> Fly is a good choice, I can imagine that it isn't very clear that "FPS" means "first-person shooter". "lightDir" technically is better, since that's what it does, it just happens that Filament only supports one directional light, which is mostly conveniently the sun. I think IBL is better because it specifically rotates the IBL (I think "Environment" is too vague"). I also think "Rotate" is better than "Arcball", because I don't think it actually implements Arcball. (It rotates around the axis in the viewing-plane perpendicular to the movement of the mouse, which I think is different.) In this case, maybe the UI should change to "Rotate".
Changed "fps" -> "fly" and "lightDir" -> "sun"